File: elixir/robot-simulator/robot_simulator.exs
defmodule RobotSimulator do
  @directions [:north, :east, :south, :west]
  @instructions ~w(L R A)
  @type position() :: {integer, integer}
  @type robot() :: {direction :: atom, position :: {integer, integer}}

  @doc """
  Create a Robot Simulator given an initial direction and position.

  Valid directions are: :north, :east, :south, :west
  """
  @spec create(direction :: atom, position :: {integer, integer}) :: {:error, String.t()} | robot()
  def create(direction \\ :north, position \\ {0, 0})
  def create(direction, _position) when direction not in @directions do
    {:error, "invalid direction"}
  end
  def create(direction, {x, y} = position) when is_integer(x) and is_integer(y) do
    {direction, position}
  end
  def create(_direction, _) do
    {:error, "invalid position"}
  end


  @doc """
  Simulate the robot's movement given a string of instructions.

  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot :: robot(), instructions :: String.t()) :: any
  def simulate(robot, instructions) when is_binary(instructions) do
    simulate(robot, String.codepoints(instructions))
  end
  def simulate(robot, []) do
    robot
  end
  def simulate(robot, [inst | instructions]) when inst in @instructions do
    {direction, position} = robot
    index = Enum.find_index(@directions, &(&1 === direction))
    robot = case inst do
      "L" ->
        {Enum.fetch!(@directions, index - 1), position}
      "R" ->
        {Enum.fetch!(Stream.cycle(@directions), index + 1), position}
      "A" ->
        {x, y} = position
        position = case direction do
          :north -> {x, y + 1}
          :east -> {x + 1, y}
          :south -> {x, y - 1}
          :west -> {x - 1, y}
        end
        {direction, position}
    end
    simulate(robot, instructions)
  end
  def simulate(_robot, _instructions) do
    {:error, "invalid instruction"}
  end

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot :: robot()) :: atom
  def direction({direction, _position}) do
    direction
  end

  @doc """
  Return the robot's position.
  """
  @spec position(robot :: robot()) :: {integer, integer}
  def position({_direction, position}) do
    position
  end
end

File: erlang/robot-simulator/src/robot_simulator.erl
-module(robot_simulator).
-behaviour(gen_statem).

-export([init/1, callback_mode/0, start_link/1, handle_event/4]).
-export([advance/1, create/0, direction/1, left/1, place/3, position/1, right/1]).

-define(COMPASS, [north, east, south, west, north]).

init(_Args) ->
  {ok, north, {0, 0}}.

callback_mode() ->
    handle_event_function.

start_link(Args) ->
  {ok, PID} = gen_statem:start_link(?MODULE, Args, []),
  PID.

handle_event(cast, {place, {Direction, Position}}, _Direction, _Position) ->
  {next_state, Direction, Position};

handle_event(cast, advance, Direction, Position) ->
  {keep_state, advance_coordinates(Direction, Position)};

handle_event(cast, Way, Direction, Position) when Way == left orelse Way == right ->
  Direction1 = turn_compass(Way, Direction, ?COMPASS),
  {next_state, Direction1, Position};

handle_event({call, From}, position, _Direction, Position) ->
  {keep_state_and_data, {reply, From, Position}};
handle_event({call, From}, direction, Direction, _Position) ->
  {keep_state_and_data, {reply, From, Direction}}.

turn_compass(left, Current, Compass) -> turn_compass(right, Current, lists:reverse(Compass));
turn_compass(_Way, Current, [Current | T]) -> hd(T);
turn_compass(Way, Current, [_ | T]) -> turn_compass(Way, Current, T).

advance_coordinates(north, {X, Y}) -> {X, Y + 1};
advance_coordinates(south, {X, Y}) -> {X, Y - 1};
advance_coordinates(east, {X, Y}) -> {X + 1, Y};
advance_coordinates(west, {X, Y}) -> {X - 1, Y}.

create() -> start_link(nil).
place(Robot, Direction, Position) -> gen_statem:cast(Robot, {place, {Direction, Position}}).
advance(Robot) -> gen_statem:cast(Robot, advance).
left(Robot) -> gen_statem:cast(Robot, left).
right(Robot) -> gen_statem:cast(Robot, right).
position(Robot) -> gen_statem:call(Robot, position).
direction(Robot) -> gen_statem:call(Robot, direction).
